### DISCUSSION TITLE
What exactly does `PreserveReferences` do to compare objects?

### DISCUSSION THREAD

**Discussion by OronDF343** [LuckyPennySoftware/AutoMapper#3939] [NA]
My scenario (simplified):
```
public class MessageViewModel
{
    public List<RecipientViewModel> EmailRecipients { get; set; }
    public List<RecipientViewModel> SmsRecipients { get; set; }
}

public class Message
{
    public List<Recipient> EmailRecipients { get; set; }
    public List<Recipient> SmsRecipients { get; set; }
}
```

The goal is to map from `MessageViewModel` to `Message`.
`EmailRecipients` and `SmsRecipients` may contain items that are in fact the same reference.
`Recipient` implements `Equals` and `GetHashCode`, while `RecipientViewModel` does not.
The documentation says almost nothing about `PreserveReferences`.

1. I should call `PreserveReferences()` on my mapping from `RecipientViewModel` to `Recipient`, correct?
2. #3654 implies that `PreserveReferences` uses `GetHashCode`.
    Does this mean that true reference equality is preserved only if the **source** type does **not** implement `GetHashCode`?
3. What happens if I reverse the mapping?

**Comment by lbargaoanu** [comment]
I suggest that you write some tests to answer your questions. You can also check our tests for `PreserveReferences`.
Also, check [the execution plan](https://docs.automapper.org/en/latest/Understanding-your-mapping.html).
